### PR TITLE
Make `Adadelta`,`Adagrad` & `Adamax` differentiable

### PR DESCRIPTION
**Pull request by emcastillo**
Continuing the differentiable optimizers support

### CODE REVIEW

File: torch/optim/adamax.py [hunk 1]
```diff
@@ -228,7 +235,10 @@ def _single_tensor_adamax(params: List[Tensor],
             exp_inf.mul_(beta2).unsqueeze(0),
             grad.abs().add_(eps).unsqueeze_(0)
         ], 0)
-        torch.amax(norm_buf, 0, keepdim=False, out=exp_inf)
+        if not differentiable:
+            torch.amax(norm_buf, 0, keepdim=False, out=exp_inf)
+        else:
+            exp_inf = torch.amax(norm_buf, 0, keepdim=False)
```
Comment (janeyx99) [inline torch/optim/adamax.py:241]: Is the distinction here to maintain functional-ism/no mutations?

**Comment by janeyx99** [comment]
@emcastillo this change looks good overall! Could you sign the CLA + rebase your PR to make the pr_sanity_check green?

The XLA failure looks unrelated, so rerunning to make sure that's the case.

File: torch/optim/adamax.py (hunk 1 above)
Comment (emcastillo) [reply torch/optim/adamax.py:241]: This is to avoid an in-place modification of `exp_inf` that will trigger autograd errors :)

**Comment by emcastillo** [comment]
Yeah, I am waiting for my company to fill the corporate CLA to fill it, so it will take a few days :)

**Comment by janeyx99** [comment]
could you rebase locally to see if you could clear the other two CI failures? you can also comment “@pytorchbot rebase” to rebase onto our stable viable/strict branch

**Comment by janeyx99** [comment]
Nice! This PR would be good to go once your company gets the CLA done.

**Review by janeyx99 — APPROVED** [review]

**Pull request closed without merging**